hathor/p2p/resources/netfilter.py

```diff
@@ -0,0 +1,402 @@
+    NetfilterMatchPeerId,
+)
+from hathor.p2p.netfilter.rule import NetfilterRule
+from hathor.p2p.netfilter.targets import NetfilterAccept, NetfilterJump, NetfilterLog, NetfilterReject
+from hathor.util import json_dumpb, json_loadb
+
+if TYPE_CHECKING:
+    from twisted.web.http import Request
+
+    from hathor.manager import HathorManager
+
+
+def handle_request(request: 'Request') -> Dict[str, Any]:
+    """ Auxiliar method to be used by POST and DELETE requests
+        to handle the parameters validation
+    """
+    if request.content is None:
+        return {'success': False, 'message': 'No body data'}
+
+    raw_data = request.content.read()
+    if raw_data is None:
+        return {'success': False, 'message': 'No body data'}
+
+    try:
+        data = json_loadb(raw_data)
+    except (JSONDecodeError, AttributeError):
+        return {'success': False, 'message': 'Invalid format for body data'}
+
+    # Map of available matches and targets with their classes
+    matches = {
+        'all': NetfilterMatchAll,
+        'and': NetfilterMatchAnd,
+        'or': NetfilterMatchOr,
+        'ip': NetfilterMatchIPAddress,
+        'peer_id': NetfilterMatchPeerId,
+    }
+
+    targets = {
+        'accept': NetfilterAccept,
+        'reject': NetfilterReject,
+        'jump': NetfilterJump,
+        'log': NetfilterLog,
+    }
+
+    # First we get the filter table chain
+    try:
+        chain = get_table('filter').get_chain(data.get('chain'))
+    except KeyError:
+        return {'success': False, 'message': 'Invalid netfilter chain.'}
+
+    # Then we get the match
+    match_value = data.get('match')
+    match_params = data.get('match_params', {})
+
+    if match_value not in matches:
+        return {'success': False, 'message': 'Invalid netfilter match.'}
+
+    match_class = matches[match_value]
+
+    try:
+        match = match_class(**match_params)
+    except TypeError:
+        return {'success': False, 'message': 'Invalid netfilter match parameters.'}
+
+    # Finally we get the target
+    target_value = data.get('target')
+    target_params = data.get('target_params', {})
+
+    if target_value not in targets:
+        return {'success': False, 'message': 'Invalid netfilter target.'}
+
+    target_class = targets[target_value]
+
+    try:
+        target = target_class(**target_params)
+    except TypeError:
+        return {'success': False, 'message': 'Invalid netfilter target parameters.'}
+
+    rule = NetfilterRule(match, target)
+
+    return {'success': True, 'chain': chain, 'rule': rule}
+
+
+@register_resource
+class NetfilterRuleResource(Resource):
+    """ Implements a web server API for handling netfilter.
+
+    POST to add a rule, GET to get rules and DELETE to remove a rule.
+
+    You must run with option `--status <PORT>`.
+    """
+    isLeaf = True
+
+    def __init__(self, manager: 'HathorManager'):
+        self.manager = manager
+
+    def render_GET(self, request: 'Request') -> bytes:
```

Comment:
If one saves the return of the GET method and send it to the POST method, would it work?

That's a common task to dump the current rules and restore them when the full node restarts.

Reply:
Changed how we expect the POST format, so we can reuse the GET response on it. Can you see again, please?